File: baekjoon/10022022/src/Problem11444.java
import java.io.BufferedReader;
import java.io.InputStreamReader;

class Main {
    private static final long MOD = 1000000007;

    private static long[][] multiply(long[][] a, long[][] b) {
        int x = a.length, y = b[0].length, z = a[0].length;
        long[][] result = new long[x][y];
        for (int i = 0; i < x; i++) {
            for (int j = 0; j < y; j++) {
                for (int k = 0; k < z; k++) {
                    result[i][j] += a[i][k] * b[k][j];
                }
                result[i][j] %= MOD;
            }
        }
        return result;
    }

    static long[][] dq(long[][] matrix, long n) {
        if (n == 1) return matrix;
        else {
            long[][] divide = dq(matrix, n/2);
            if (n % 2 == 1) return multiply(multiply(divide, divide), matrix);
            else return multiply(divide, divide);
        }
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        long n = Long.parseLong(br.readLine());
        long[][] init = {{0, 1}, {1, 1}};
        System.out.print(dq(init, n)[0][1]);
    }
}

File: baekjoon/10022022/src/Problem11444.kt
import java.io.StreamTokenizer

private const val mod = 1000000007

fun main() = with(StreamTokenizer(System.`in`.bufferedReader())) {
    nextToken(); val n = nval.toLong()

    //  F(n+1) Fn
    //  Fn  F(n-1)  형태로 형렬을 만들고,
    //  해당 행렬을 제곱하면서 피보나치 수열의 값을 만들 수 있다.
    val init = Array(2) { Array(2) { 1L } }
    init[1][1] = 0

    print(dq(init, n)[0][1])
}

private fun dq(matrix: Array<Array<Long>>, exponent: Long): Array<Array<Long>> {
    return if (exponent == 1L) matrix
    else {
        val divide = dq(matrix, exponent/2)
        if (exponent % 2 == 0L) multiply(divide, divide)
        else multiply(multiply(divide, divide), matrix)
    }
}

private fun multiply(a: Array<Array<Long>>, b: Array<Array<Long>>): Array<Array<Long>> {
    val result = Array(a.size) { Array(b.size) { 0L } }
    for (i: Int in a.indices) {
        for (j: Int in b[i].indices) {
            for (k: Int in b.indices) {
                result[i][j] += a[i][k] * b[k][j]
            }
            result[i][j] = result[i][j] % mod
        }
    }
    return result
}